### PR TITLE
Add linux conntrack metrics to host metrics receiver

### PR DESCRIPTION
**Pull request by emilgelman**
**Description:**
Added support for collecting conntrack metrics (linux only) using the host metrics receiver. The metrics are taken from https://github.com/shirou/gopsutil (already being used in the receiver)

### CODE REVIEW

**Comment by emilgelman** [comment]
@mx-psi
Reopened in a new PR to fix the EasyCLA issues.
Please review

File: receiver/hostmetricsreceiver/internal/scraper/networkscraper/factory.go
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/networkscraper/factory.go:39]: We can/should rely on the zero value for booleans being `false, so this should also work 
```suggestion
	return &Config{}
```

File: receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_linux.go
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_linux.go:46]: Add context to the error
```suggestion
		return fmt.Errorf("failed to scrape conntrack metrics: %w", err)
```

File: receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_others.go
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_others.go:39]: Should we instead return an error to indicate that this is unsupported on other platforms?

File: receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_scraper.go
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_scraper.go:54]: Should we use ` FilterCountersWithContext` instead, so that we can pass the context in case we want to have cancellation at some point? 

Not super convinced that we should do this here (`IOCounters` also has `IOCountersWithContext` and we are not using it), so maybe you can just open an issue and we can deal with it in a separate PR.

File: receiver/hostmetricsreceiver/metadata.yaml
Comment (mx-psi) [inline receiver/hostmetricsreceiver/metadata.yaml:298]: Something like
```suggestion
    unit: "{flow entries}"
```
might be more useful

File: receiver/hostmetricsreceiver/metadata.yaml
Comment (mx-psi) [inline receiver/hostmetricsreceiver/metadata.yaml:306]: Same comment as above
```suggestion
    unit: "{flow entries}"
```

File: receiver/hostmetricsreceiver/metadata.yaml
Comment (mx-psi) [inline receiver/hostmetricsreceiver/metadata.yaml:296]: I am not knowledgeable enough about macOS or Windows to know if there are equivalents to conntrack in these OSs. It might be a good idea to check this and make the name more general in case we want to add support for other platforms later.

A good way to get feedback on this is to open a PR on the https://github.com/open-telemetry/opentelemetry-specification repository so that other people can chime in.

File: receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_scraper_test.go
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_scraper_test.go:167]: For the test failure, we can just mock the `conntrack` call if it's hard to make this work properly on the `coverage` image. From a quick look I am not sure why it doesn't work since, as you said, the necessary file is there.

Mocking also provides a good oportunity to add a test for when `conntrack` returns an error.

**Review by mx-psi — COMMENTED** [review]
I left some comments :)

File: receiver/hostmetricsreceiver/README.md
Comment (emilgelman) [reply receiver/hostmetricsreceiver/README.md:74]: I'm not sure what's the best option as well, especially considering the fact that I wasn't able to find any alternative of conntrack for Windows/ macOS. Is a seperate scraper a cleaner solution in case this is only relevant for Linux?

Comment (mx-psi) [reply receiver/hostmetricsreceiver/README.md:74]: The `processes`scraper seems to be Linux only so `conntrack` wouldn't be the first scraper lacking cross-platform support

**Comment by mx-psi** [comment]
Unit tests are failing with

```
--- FAIL: TestGatherMetrics_EndToEnd (0.17s)
    hostmetrics_receiver_test.go:177: 
        	Error Trace:	hostmetrics_receiver_test.go:177
        	            				hostmetrics_receiver_test.go:152
        	            				asm_amd64.s:1371
        	Error:      	Not equal: 
        	            	expected: 26
        	            	actual  : 24
        	Test:       	TestGatherMetrics_EndToEnd
    hostmetrics_receiver_test.go:179: 
        	Error Trace:	hostmetrics_receiver_test.go:179
        	            				hostmetrics_receiver_test.go:152
        	            				asm_amd64.s:1371
        	Error:      	map[string]struct {}{"system.cpu.load_average.15m":struct {}{}, "system.cpu.load_average.1m":struct {}{}, "system.cpu.load_average.5m":struct {}{}, "system.cpu.time":struct {}{}, "system.disk.io":struct {}{}, "system.disk.io_time":struct {}{}, "system.disk.merged":struct {}{}, "system.disk.operation_time":struct {}{}, "system.disk.operations":struct {}{}, "system.disk.pending_operations":struct {}{}, "system.disk.weighted_io_time":struct {}{}, "system.filesystem.inodes.usage":struct {}{}, "system.filesystem.usage":struct {}{}, "system.memory.usage":struct {}{}, "system.network.connections":struct {}{}, "system.network.dropped":struct {}{}, "system.network.errors":struct {}{}, "system.network.io":struct {}{}, "system.network.packets":struct {}{}, "system.paging.faults":struct {}{}, "system.paging.operations":struct {}{}, "system.paging.usage":struct {}{}, "system.processes.count":struct {}{}, "system.processes.created":struct {}{}} does not contain "system.conntrack.count"
        	Test:       	TestGatherMetrics_EndToEnd
    hostmetrics_receiver_test.go:179: 
        	Error Trace:	hostmetrics_receiver_test.go:179
        	            				hostmetrics_receiver_test.go:152
        	            				asm_amd64.s:1371
        	Error:      	map[string]struct {}{"system.cpu.load_average.15m":struct {}{}, "system.cpu.load_average.1m":struct {}{}, "system.cpu.load_average.5m":struct {}{}, "system.cpu.time":struct {}{}, "system.disk.io":struct {}{}, "system.disk.io_time":struct {}{}, "system.disk.merged":struct {}{}, "system.disk.operation_time":struct {}{}, "system.disk.operations":struct {}{}, "system.disk.pending_operations":struct {}{}, "system.disk.weighted_io_time":struct {}{}, "system.filesystem.inodes.usage":struct {}{}, "system.filesystem.usage":struct {}{}, "system.memory.usage":struct {}{}, "system.network.connections":struct {}{}, "system.network.dropped":struct {}{}, "system.network.errors":struct {}{}, "system.network.io":struct {}{}, "system.network.packets":struct {}{}, "system.paging.faults":struct {}{}, "system.paging.operations":struct {}{}, "system.paging.usage":struct {}{}, "system.processes.count":struct {}{}, "system.processes.created":struct {}{}} does not contain "system.conntrack.max"
        	Test:       	TestGatherMetrics_EndToEnd
FAIL
```

**Comment by emilgelman** [comment]
@mx-psi First of all, thank you for the review.
I've also opened a PR in the specification repo:
https://github.com/open-telemetry/opentelemetry-specification/pull/1761

Still trying to understand why the e2e test fails, running the same docker image with this branch seems to work fine.

**Comment by emilgelman** [comment]
@mx-psi Please review again.
I wasn't able to validate the new conntrack metrics in `TestGatherMetrics_EndToEnd`
I tried reproducing the CI flow using cimg/go:1.16, and it does work locally. However, it seems like the netfilter files are missing when executed through the CI:
https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector/8159/workflows/4cca0725-3807-4123-821d-17a5faa63eb9/jobs/88360

```
        	Error:      	Received unexpected error:
        	            	open /proc/sys/net/netfilter/nf_conntrack_count: no such file or directory
```

Although they exist:
```
➜  ~ docker run  cimg/go:1.16 ls -la  /proc/sys/net/netfilter/nf_conntrack_count

-r--r--r-- 1 root root 0 Jun 16 20:08 /proc/sys/net/netfilter/nf_conntrack_count
➜  ~ docker run  cimg/go:1.16 ls -la  /proc/sys/net/netfilter/nf_conntrack_max

-rw-r--r-- 1 root root 0 Jun 16 20:08 /proc/sys/net/netfilter/nf_conntrack_max
```

File: receiver/hostmetricsreceiver/README.md
Comment (mx-psi) [inline receiver/hostmetricsreceiver/README.md:36]:
```suggestion
| conntrack  | Linux  <sup>[2]</sup>         | Conntrack count and max connections metrics            |
```

And then add a note with details about what setup is needed for installation

**Review by mx-psi — COMMENTED** [review]
Looks good, just one documentation comment! I think it's fine to not have the conntrack end to end test (but maintainers may disagree with me)

**Review by mx-psi — APPROVED** [review]
Looks good to me! Now an approver/maintainer should have a look; remember that the focus is on getting to GA so please be patient; if this doesn't get attention feel free to raise this on the CNCF Slack

**Comment by mx-psi** [comment]
Not stale but I don't have permissions to remove the label (I hope this comment is enough to mark it as active)

**Comment by albertvaka** [comment]
Hi! I'm commenting to remove the stale label. mx-psi is on vacation but will be reviewing this when he's back :)

**Comment by mx-psi** [comment]
@emilgelman can you update the PR with the new `Number` types as was done in #3710?

File: receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_test.go
Comment (codeboten) [inline receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_test.go:91]:
```suggestion
	assert.Equal(t, 1, metric.Sum().DataPoints().Len())
```

File: receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_linux.go
Comment (codeboten) [inline receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_linux.go:49]:
```suggestion
	idps := metric.Sum().DataPoints()
```

File: receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_linux.go
Comment (codeboten) [inline receiver/hostmetricsreceiver/internal/scraper/conntrackscraper/conntrack_scraper_linux.go:56]:
```suggestion
func initializeNetworkConntrackDataPoint(dataPoint pdata.NumberDataPoint, now pdata.Timestamp, value int64) {
	dataPoint.SetTimestamp(now)
	dataPoint.SetIntVal(value)
```

File: receiver/hostmetricsreceiver/metadata.yaml
Comment (codeboten) [inline receiver/hostmetricsreceiver/metadata.yaml:300]:
```suggestion
      type: sum
```

File: receiver/hostmetricsreceiver/metadata.yaml
Comment (codeboten) [inline receiver/hostmetricsreceiver/metadata.yaml:308]:
```suggestion
      type: sum
```

File: receiver/hostmetricsreceiver/internal/metadata/generated_metrics.go
Comment (codeboten) [inline receiver/hostmetricsreceiver/internal/metadata/generated_metrics.go:59]: This will need to be regenerated after the suggestions are merged to change from `IntSum` to `Sum`

**Review by codeboten — COMMENTED** [review]
The code looks good, I added some suggestions to change the code from IntSum to Sum as per @mx-psi's recommendation, please take a look.

**Comment by bogdandrutu** [comment]
@emilgelman sorry, unfortunately this needs to be moved to contrib, since we moved the hostmetricsreceiver there.

**Pull request closed without merging**
